### DISCUSSION TITLE
Regarding usage of `UTILS_LIKELY`/`UTILS_UNLIKELY`

### DISCUSSION THREAD

**Discussion by adamwych** [google/filament#7959] [Q&A] [answered]
Hello. When browsing through the code I noticed a lot of `UTILS_LIKELY` and `UTILS_UNLIKELY` calls near branches. As I understood (from the [docs](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fexpect)), they tell the compiler whether a branch is... likely or unlikely - but often those calls seem to be inverted.

For example, here, when creating a material instance (https://github.com/google/filament/blob/main/filament/src/details/Engine.cpp#L808):

```cpp
FMaterialInstance* FEngine::createMaterialInstance(const FMaterial* material,
        const FMaterialInstance* other, const char* name) noexcept {
    FMaterialInstance* p = mHeapAllocator.make<FMaterialInstance>(*this, other, name);
    if (UTILS_UNLIKELY(p)) { // should never happen
        auto pos = mMaterialInstances.emplace(material, "MaterialInstance");
        pos.first->second.insert(p);
    }
    return p;
}
```

This `if` would evaluate to something like:
```cpp
if (__builtin_expect (p, false)) {
  // ...
}
```

The branch here is marked as unlikely (as in, we expect it to be `0`), but I think it should be the opposite - it's expected to happen basically always? Did I understand it incorrectly or is this a mistake?

@pixelflinger I saw that your commit added it, can you explain?

**Comment by pixelflinger** [comment, marked as answer]
Oh that definitely looks like a mistake.
